esitoolsupport/esibase.py

```diff
@@ -0,0 +1,114 @@
+# the Free Software Foundation; version 3 of the License.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see http://www.gnu.org/licenses/.
+#
+# Please contact Eucalyptus Systems, Inc., 6755 Hollister Ave., Goleta
+# CA 93117, USA or visit http://www.eucalyptus.com/licenses/ if you need
+# additional information or have any questions.
+
+#
+# Order matters here. We want to make sure we initialize logging before anything
+# else happens. We need to initialize the logger that boto will be using.
+#
+
+import sys
+import subprocess
+import os
+from urlparse import urlparse
+
+
+class EsiBase(object):
+
+    def __init__(self, region='localhost'):
+        self.vars = {}
+        self.region = region
+        self._load_vars()
+        self.check_environment()
+
+    def list_system_accounts(self):
+        accounts = {}
+        process = subprocess.Popen(['/usr/bin/euare-accountlist',
+                                    '-U', self.vars["AWS_IAM_URL"],
+                                    '-I', self.vars["AWS_ACCESS_KEY_ID"],
+                                    '-S', self.vars["AWS_SECRET_ACCESS_KEY"]],
+                                   stdout=subprocess.PIPE)
+        t = process.communicate()
+        if process.returncode == 0:
+            for account in t[0].split('\n'):
+                a = account.split('\t')
+                if len(a) == 2:
+                    accounts[a[0]] = a[1]
+        return accounts
+
+    def _load_vars(self):
+        EsiBase._check_binary(['euca-generate-environment-config'])
+        process = subprocess.Popen(['euca-generate-environment-config', '--simple', '--region',
+                                    self.region], stdout=subprocess.PIPE)
+        t = process.communicate()
+        if process.returncode == 0:
+            for var in t[0].split('\n'):
+                v = var.split("=")
+                if len(v) == 2:
+                    self.vars[v[0]] = v[1] if v[1] != '' else None
+
+    @staticmethod
+    def _check_binary(binary):
+        try:
+            with open(os.devnull, 'w') as nullfile:
+                subprocess.call(binary, env=os.environ.copy(), stdout=nullfile)
+        except OSError:
+            print >> sys.stderr, "Error: cannot execute '{0}' binary.".format(" ".join(binary))
+            print >> sys.stderr, "Make sure EUCALYPTUS path variable is exported."
+            sys.exit(1)
+
+    def _get_properties_url(self):
+        url = urlparse(self.vars["EC2_URL"])
+        port = url.port if url.port else 80
+        return "{0}://{1}:{2}{3}".format(url.scheme, url.hostname, port, '/services/Properties')
```

Comment:
If EC2_URL is DNS-based this will result in a URL that doesn't work because it says "compute" in its DNS name and "/services/Properties" in its service path.  euca-generate-environment-config will output EUCA_PROPERTIES_URL if it knows of one, so consider using that.